File: 328-OddEvenLinkedList-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (head==NULL||head->next==NULL) return head;
        
        ListNode* p1=head, *p2=head->next;
        while (p2!=NULL && p2->next!=NULL) {
            ListNode* temp = p2->next;
            p2->next=temp->next;
            temp->next=p1->next;
            p1->next=temp;
            p1=temp;
            p2=p2->next;
        }
        
        return head;
        
    }
};

File: 328-OddEvenLinkedList-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head==null||head.next==null) return head;
        ListNode head1=head,head2=head.next;
        while (head1!=null&&head2.next!=null) {
            ListNode head3=head2.next;
            head2.next=head3.next;
            head3.next=head1.next;
            head1.next=head3;
            head1=head3;
            head2=head2.next;
            if (head2==null) break;
        }
        return head;
    }
}
